### PR TITLE
FFI: Add possibility to load ECDSA and ECDH keys from MPI

### LINKED ISSUE #1073: FFI: Add possibility to load key pair for ECDSA sign/verification
It doesn't seem that FFI supports loading of EC keys in format different than BER. We need functions similar to botan_pubkey_load_dsa/botan_prvkey_load_dsa. 

I'll create a patch today

### PR DESCRIPTION
**Pull request by flowher**
This patch resolves #1073

This work was sponsored by Ribose Inc

### CODE REVIEW

File: src/lib/ffi/ffi.cpp [hunk 1]
```diff
@@ -1699,6 +1757,64 @@ int botan_pubkey_ed25519_get_pubkey(botan_pubkey_t key,
 #endif
    }
 
+int botan_pubkey_load_ecdsa(botan_pubkey_t* key,
+                                   const botan_mp_t public_x,
+                                   const botan_mp_t public_y,
+                                   const char* curve_name)
+   {
+
+   std::unique_ptr<Botan::ECDSA_PublicKey> p_key;
```
Comment (randombit) [inline src/lib/ffi/ffi.cpp]: This will fail to compile if ECDSA isn't enabled

File: src/lib/ffi/ffi.cpp [hunk 2]
```diff
@@ -1699,6 +1757,64 @@ int botan_pubkey_ed25519_get_pubkey(botan_pubkey_t key,
...
+
+   return -1;
+   }
+
+int botan_pubkey_load_ecdh(botan_pubkey_t* key,
+                                   const botan_mp_t public_x,
+                                   const botan_mp_t public_y,
+                                   const char* curve_name)
+   {
+   std::unique_ptr<Botan::ECDH_PublicKey> p_key;
```
Comment (randombit) [inline src/lib/ffi/ffi.cpp]: Ditto for ECDH

File: src/lib/ffi/ffi.cpp [hunk 3]
```diff
@@ -330,8 +330,66 @@ Botan::BigInt privkey_get_field(const Botan::Private_Key& key,
    throw Botan::Exception("Unsupported algorithm type for botan_privkey_get_field");
    }
 
+template<class ECPrivateKey_t>
+int privkey_load_ec(std::unique_ptr<ECPrivateKey_t>& key,
```
Comment (randombit) [inline src/lib/ffi/ffi.cpp]: These functions are not exposed so should be in an anon namespace

File: src/lib/ffi/ffi.cpp [hunk 4]
```diff
@@ -1699,6 +1757,64 @@ int botan_pubkey_ed25519_get_pubkey(botan_pubkey_t key,
...
+
+   return -1;
+   }
+
+int botan_privkey_load_ecdsa(botan_privkey_t* key,
+                          const botan_mp_t scalar,
+                          const char* curve_name)
+   {
+   std::unique_ptr<Botan::ECDSA_PrivateKey> p_key;
+   if(!privkey_load_ec(p_key, safe_get(scalar), curve_name))
```
Comment (randombit) [inline src/lib/ffi/ffi.cpp]: If the user passes something that isn't a `botan_mp_t` for the scalar, `safe_get` will throw. Instead pass the `botan_mp_t` to the helper and dereference it in the `try`/`catch` block

**Comment by flowher** [comment]
Changes applied. 
It seems to me that travis has some problems with it's environment.

**Comment by randombit** [comment]
I could restart the jobs but looks certainly to be a Travis issue so I'm going to go ahead and merge. Thanks!

**Comment by flowher** [comment]
Cool, thanks!

**Pull request merged**